lib/web_ui/lib/src/engine/text/paragraph.dart

```diff
@@ -189,6 +104,117 @@ class EngineLineMetrics implements ui.LineMetrics {
   }
 }
 
+class RichEngineLineMetrics extends EngineLineMetrics {
```

Comment:
The sample you provided could work, but it means I have to add `!` everywhere:
```dart
lineMetrics.richMetrics!.*
```
Reversing the composition might work better in this case:
```dart
class ParagraphLine { // Instead of RichEngineLineMetrics
  const ParagraphLine(this.lineMetrics, ... other properties);

  final ui.LineMetrics lineMetrics;
}
```